### DISCUSSION TITLE
Programatically CPU/Network throttling  Cypress test

### DISCUSSION THREAD

**Discussion by thevladisss** [cypress-io/cypress#29403] [Questions and Help]
Hi. I've been thinking of a possibility of programatically forcing Cypress test to set specified CPU or Network conditions. Is that now possible somehow besides that it is mentioned in the docs that it can be done manually?

_i was going to make a feature issue, however, thought this should go first_

Thanks

**Comment by portyu9** [comment]
Hi @thevladisss 
Yes, this is possible in Chromium-based browsers through the Chrome DevTools Protocol, although Cypress does not currently expose a supported cross-browser throttling API.

For example:

```js
const cdp = (command, params = {}) =>
  cy.wrap(
    Cypress.automation('remote:debugger:protocol', {
      command,
      params,
    }),
    { log: false }
  )

// 4x CPU slowdown
cdp('Emulation.setCPUThrottlingRate', {
  rate: 4,
})

// Example network throttling
cdp('Network.emulateNetworkConditions', {
  offline: false,
  latency: 200,
  downloadThroughput: (1 * 1024 * 1024) / 8,
  uploadThroughput: (500 * 1024) / 8,
})
```

Reset it after the test:

```js
cdp('Emulation.setCPUThrottlingRate', {
  rate: 1,
})

cdp('Network.emulateNetworkConditions', {
  offline: false,
  latency: 0,
  downloadThroughput: -1,
  uploadThroughput: -1,
})
```

The important caveat is that `remote:debugger:protocol` is an internal/undocumented Cypress API and only applies to Chromium-family browsers, so I would not build portable test infrastructure around it.

For a supported cross-browser Cypress feature, there currently isn't an equivalent API for programmatically setting CPU/network conditions.

**Reply by portyu9** [reply]
@thevladisss Please mark this answer as accepted if it helped, thank you.
